### DISCUSSION TITLE
Docs or demos showing use of SQL Server data in RAG applications?

### DISCUSSION THREAD

**Discussion by JasonHaley** [microsoft/AI-Chat-App-Hack#31] [General]
Does anyone know where I can find some docs or demos showing use of SQL Server data in RAG applications?

In my work life, I work with a lot of business applications where almost all the data is in sql server databases (not documents). Does anyone know of a good resource for learning best practices to add business data from sql databases into a RAG system.

My guess is I'll need to identify the useful columns, then get some embeddings and keep the vectors for using in the initial search phase - but I'd love to find some useful examples to see how people are doing it already.

I'm also guessing I'll need some sort of index updater as data can change in some systems on a regular basis.

**Comment by pamelafox** [comment]
Hi @JasonHaley, great questions. Here are a few related notebooks:

SQL DB ingestion -
https://github.com/sdt-msft/Azure-Cognitive-Search-Azure-OpenAI-Accelerator/blob/main/07-SQLDB_QA.ipynb

Tabular data generally - 
https://github.com/sdt-msft/Azure-Cognitive-Search-Azure-OpenAI-Accelerator/blob/main/06-TabularDataQA.ipynb

I think the big question whenever you're RAGing on a database is whether you are going to directly issue the search query to the database, or if you're going to first ingest the database into an Azure AI search index.

Benefits of direct query - works well for data that's changing a lot, where indexing could be annoying/costly/error-prone, such as retail inventory.

Drawbacks of direct query - database must support vector search, and ideally also support keyword search, and some form of hybrid search (keyword + vector), for optimal relevance in results. Azure AI search offers all that built-in.

I also found additional tutorials when I searched for "sqlserver rag openai" on the web. I think there may be an ingester for Azure AI Search - you could ask in tomorrow's session with the search team.

**Comment by JasonHaley** [comment]
This weekend I got this example up and running - it has a SQL Server with business data in it: [https://github.com/Azure-Samples/openai/tree/main/End_to_end_Solutions/AOAISearchDemo](https://github.com/Azure-Samples/openai/tree/main/End_to_end_Solutions/AOAISearchDemo)
